include/mbedtls/check_config.h

```diff
@@ -66,6 +66,71 @@
 #error "MBEDTLS_HAVE_TIME_DATE without MBEDTLS_HAVE_TIME does not make sense"
 #endif
 
+/* Check that each MBEDTLS_ECP_DP_xxx symbol has its PSA_WANT_ECC_xxx counterpart,
+ * but only when config_psa.h is evaluated. */
+#if defined(MBEDTLS_PSA_CRYPTO_CONFIG) || defined(MBEDTLS_PSA_CRYPTO_C)
+
+#if defined(MBEDTLS_ECP_DP_BP256R1_ENABLED) && !defined(PSA_WANT_ECC_BRAINPOOL_P_R1_256)
+#error "MBEDTLS_ECP_DP_BP256R1_ENABLED defined, but not its PSA counterpart"
+#endif
+
+#if defined(MBEDTLS_ECP_DP_BP384R1_ENABLED) && !defined(PSA_WANT_ECC_BRAINPOOL_P_R1_384)
+#error "MBEDTLS_ECP_DP_BP384R1_ENABLED defined, but not its PSA counterpart"
+#endif
+
+#if defined(MBEDTLS_ECP_DP_BP512R1_ENABLED) && !defined(PSA_WANT_ECC_BRAINPOOL_P_R1_512)
+#error "MBEDTLS_ECP_DP_BP512R1_ENABLED defined, but not its PSA counterpart"
+#endif
+
+#if defined(MBEDTLS_ECP_DP_CURVE25519_ENABLED) && !defined(PSA_WANT_ECC_MONTGOMERY_255)
+#error "MBEDTLS_ECP_DP_CURVE25519_ENABLED defined, but not its PSA counterpart"
+#endif
+
+#if defined(MBEDTLS_ECP_DP_CURVE448_ENABLED) && !defined(PSA_WANT_ECC_MONTGOMERY_448)
+#error "MBEDTLS_ECP_DP_CURVE448_ENABLED defined, but not its PSA counterpart"
+#endif
+
+#if defined(MBEDTLS_ECP_DP_SECP192R1_ENABLED) && !defined(PSA_WANT_ECC_SECP_R1_192)
+#error "MBEDTLS_ECP_DP_SECP192R1_ENABLED defined, but not its PSA counterpart"
+#endif
+
+#if defined(MBEDTLS_ECP_DP_SECP224R1_ENABLED) && !defined(PSA_WANT_ECC_SECP_R1_224)
+#error "MBEDTLS_ECP_DP_SECP224R1_ENABLED defined, but not its PSA counterpart"
+#endif
+
+#if defined(MBEDTLS_ECP_DP_SECP256R1_ENABLED) && !defined(PSA_WANT_ECC_SECP_R1_256)
+#error "MBEDTLS_ECP_DP_SECP256R1_ENABLED defined, but not its PSA counterpart"
+#endif
+
+#if defined(MBEDTLS_ECP_DP_SECP384R1_ENABLED) && !defined(PSA_WANT_ECC_SECP_R1_384)
+#error "MBEDTLS_ECP_DP_SECP384R1_ENABLED defined, but not its PSA counterpart"
+#endif
+
+#if defined(MBEDTLS_ECP_DP_SECP521R1_ENABLED) && !defined(PSA_WANT_ECC_SECP_R1_521)
+#error "MBEDTLS_ECP_DP_SECP521R1_ENABLED defined, but not its PSA counterpart"
+#endif
+
+#if defined(MBEDTLS_ECP_DP_SECP192K1_ENABLED) && !defined(PSA_WANT_ECC_SECP_K1_192)
+#error "MBEDTLS_ECP_DP_SECP192K1_ENABLED defined, but not its PSA counterpart"
+#endif
+
+/* SECP224K1 is buggy in PSA API so we skip this check */
+#if 0 && defined(MBEDTLS_ECP_DP_SECP224K1_ENABLED) && !defined(PSA_WANT_ECC_SECP_K1_224)
+#error "MBEDTLS_ECP_DP_SECP224K1_ENABLED defined, but not its PSA counterpart"
+#endif
+
+#if defined(MBEDTLS_ECP_DP_SECP256K1_ENABLED) && !defined(PSA_WANT_ECC_SECP_K1_256)
+#error "MBEDTLS_ECP_DP_SECP256K1_ENABLED defined, but not its PSA counterpart"
+#endif
+
+#endif /* MBEDTLS_PSA_CRYPTO_CONFIG || MBEDTLS_PSA_CRYPTO_C */
+
+#if defined(MBEDTLS_SOME_BUILTIN_EC) && \
+    !(defined(MBEDTLS_ECDSA_C) || defined(MBEDTLS_ECDSA_DETERMINISTIC) || \
+      defined(MBEDTLS_ECDH_C) || defined(MBEDTLS_ECJPAKE_C))
+#error "Some builtin curve is included, but there is no algorighm using it"
```

Comment:
Should I replace it with a `#warning` then? Or is it better to remove the check entirely?

Another (perhaps weird) option that comes into my mind is to create a new build symbol like `MBEDTLS_CUSTOM_EC_USAGE` and add it to this list. It wouldn't be used anywhere in the lib/test/program codes though, unless we design some specific example for that...

Wdyt?

Reply:
Sorry @gilles-peskine-arm to resume this discussion again, but I think I got to a point in which I need your feedback.

I'm rebasing this PR after the merge of #8159 and I realized that previous to this rebase I was basically not using the new internal symbol `MBEDTLS_SOME_BUILTIN_EC` anywhere in the code. I was only defining it. I guess this does not satisfy Manuel's original requirement in #8016 (2nd part of the sentence after the "OR"):

> check that flag when deciding whether to enable ECDH_C, ECDSA_C and ECJPAKE_C: we'll enable the built-in implementation if there's no accelerator for this alg OR if the flag is set (which is the new part)

I come up with my proposal to add a check in `check_config.h` after [I realized](https://github.com/Mbed-TLS/mbedtls/pull/8075#issuecomment-1678825516) that I were implementing things as Manuel asked, then some test components were failing:
- `component_test_psa_crypto_config_accel_ecdsa`
- `component_test_psa_crypto_config_accel_ecdh`
- `component_test_psa_crypto_config_accel_ecc_ecp_light_only`
- `component_test_psa_crypto_config_accel_all_curves_except_p192`

Let's pick for example `component_test_psa_crypto_config_accel_ecc_ecp_light_only`. There we accelerate all EC algs, but at the same time we keep builtin curves because they are needed for `ECP_LIGHT` features like `MBEDTLS_PK_PARSE_EC_EXTENDED`. Unfortunately in that test component we check that:
```
    not grep mbedtls_ecdsa_ library/ecdsa.o
    not grep mbedtls_ecdh_ library/ecdh.o
    not grep mbedtls_ecjpake_ library/ecjpake.o
    not grep mbedtls_ecp_mul library/ecp.o
```
and this of course fails because in `config_adjust_legacy_from_psa.h` we have something like:
```
#if defined(PSA_WANT_ALG_ECDH)
#if !defined(MBEDTLS_PSA_ACCEL_ALG_ECDH) || defined(MBEDTLS_SOME_BUILTIN_EC)
#define MBEDTLS_PSA_BUILTIN_ALG_ECDH 1
#define MBEDTLS_ECDH_C
#define MBEDTLS_ECP_C
#define MBEDTLS_BIGNUM_C
#endif /* !MBEDTLS_PSA_ACCEL_ALG_ECDH */
#endif /* PSA_WANT_ALG_ECDH */
```
which brings the builtin implementation in even though we have ECDH accelerated.

How do you suggest to overcome this problem?

Reply:
This commit https://github.com/Mbed-TLS/mbedtls/pull/8075/commits/7ef50aeb3886b64f15c0d17024cf11208b1fe612 is the best compromise I can think of. It matches Manuel's requirement and at the same time should make the CI happy.
I'm not super happy with it because it partially re-uses the same symbols that are used to enable `ECP_LIGHT`. It might be more elegant to use `ECP_LIGHT` instead here.
Anyway, is this a reasonable solution to you?

Reply:
Checking again, I think you're right that 55efa0bc693ead13063609be5e497cc4ef23308d doesn't meet the requirements in #8016. But neither does d60abc8dfca9c4815678798035e56d22d8eeb6d1.

See https://github.com/Mbed-TLS/mbedtls/pull/8075#discussion_r1319836847